Node Initialization order is wrong

While trying to setup Hazelcast on AWS I got a strange NullPointerException:

```
Caused by: java.lang.NullPointerException: null
        at com.hazelcast.partition.membergroup.SPIAwareMemberGroupFactory.createInternalMemberGroups(SPIAwareMemberGroupFactory.java:59)
```

Digging a little bit lead me to the Node class, line 209. There the PartitionService is initialized with the yet-uninitialized node. A few calls up the stack, through PartitionStateManager, there's a call to node.getDiscoveryService() which at this stage returns null and hence the SPIAwareMemberGroupFactory is initialized with a null discovery service. This is rather unfortunate, as it means practically Hazelcast is not working in this setup (SPI discovery).

--- a/hazelcast/src/main/java/com/hazelcast/instance/Node.java
+++ b/hazelcast/src/main/java/com/hazelcast/instance/Node.java
@@ -213,12 +213,12 @@ public class Node {
 
             clientEngine = new ClientEngineImpl(this);
             connectionManager = nodeContext.createConnectionManager(this, serverSocketChannel);
+            discoveryService = createDiscoveryService(
+                    this.config.getNetworkConfig().getJoin().getDiscoveryConfig().getAsReadOnly(), localMember);
             partitionService = new InternalPartitionServiceImpl(this);
             clusterService = new ClusterServiceImpl(this, localMember);
             textCommandService = new TextCommandServiceImpl(this);
             multicastService = createMulticastService(addressPicker.getBindAddress(), this, config, logger);
-            discoveryService = createDiscoveryService(
-                    this.config.getNetworkConfig().getJoin().getDiscoveryConfig().getAsReadOnly(), localMember);
             joiner = nodeContext.createJoiner(this);
         } catch (Throwable e) {
             try {

--- a/hazelcast/src/test/java/com/hazelcast/spi/discovery/DiscoverySpiTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/discovery/DiscoverySpiTest.java
@@ -81,6 +81,7 @@ import java.util.Map;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicInteger;
 
+import static com.hazelcast.config.PartitionGroupConfig.MemberGroupType.SPI;
 import static com.hazelcast.config.properties.PropertyTypeConverter.BOOLEAN;
 import static com.hazelcast.spi.discovery.DiscoverySpiTest.ParametrizedDiscoveryStrategyFactory.BOOL_PROPERTY;
 import static org.junit.Assert.assertEquals;
@@ -421,6 +422,31 @@ public class DiscoverySpiTest extends HazelcastTestSupport {
             verify(discoveryService, atLeastOnce()).discoverNodes();
         } finally {
             instance.getLifecycleService().terminate();
+        }
+    }
+
+    @Test
+    public void testMemberGroup_givenSPIMemberGroupIsActived_whenInstanceStarting_wontThrowNPE() {
+        // this test has no assert. it's a regression test checking an instance can start when a SPI-driven member group
+        // strategy is configured. see #11681
+        Config config = new Config();
+        config.setProperty(GroupProperty.DISCOVERY_SPI_ENABLED.getName(), "true");
+        JoinConfig joinConfig = config.getNetworkConfig().getJoin();
+        joinConfig.getMulticastConfig().setEnabled(false);
+
+        DiscoveryStrategyConfig discoveryStrategyConfig = new DiscoveryStrategyConfig(MetadataProvidingDiscoveryStrategy.class.getName());
+        joinConfig.getDiscoveryConfig()
+                .addDiscoveryStrategyConfig(discoveryStrategyConfig);
+        config.getPartitionGroupConfig().setGroupType(SPI).setEnabled(true);
+
+        HazelcastInstance hazelcastInstance = null;
+        try {
+            // check the instance can actually be started
+            hazelcastInstance = Hazelcast.newHazelcastInstance(config);
+        } finally {
+            if (hazelcastInstance != null) {
+                hazelcastInstance.shutdown();
+            }
         }
     }
 

